### PR TITLE
Add `constness` to `TraitDef`

### PR DESCRIPTION
**Pull request by fee1-dead**
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/127200)*

Second attempt at fixing the regression @ https://github.com/rust-lang/rust/pull/120639#issuecomment-2198373716

r? project-const-traits

### CODE REVIEW

**Review by compiler-errors — COMMENTED** [review]
You should be able to set `EncodeCrossCrate::No` for `const_trait`, though I don't think it'll improve perf :/

**Comment by fee1-dead** [comment]
> You should be able to set `EncodeCrossCrate::No` for `const_trait`, though I don't think it'll improve perf :/

done

@rustbot ready

**Comment by fee1-dead** [comment]
FWIW, I think this is a perf improvement. Looking at the details of the single perf regression seems to show more instructions being spent on loading incremental data. That might be because we added `constness` to `TraitDef`. But cycles doesn't have any significant changes so I think this should be fine.

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs [hunk 1]
```diff
@@ -1847,19 +1847,13 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
                     path.segments[..path.segments.len() - 2].iter(),
                     GenericsArgsErrExtend::None,
                 );
-                // HACK: until we support `<Type as ~const Trait>`, assume all of them are.
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:1850]: Why was this removed? Do we not need this anymore?

**Review by compiler-errors — APPROVED** [review]
One question, r=me otherwise

File: compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs (hunk 1 above)
Comment (fee1-dead) [reply compiler/rustc_hir_analysis/src/hir_ty_lowering/mod.rs:1850]: since we don't have the constness param on traits, we don't need <T as ~const> anymore
